### ISSUE TITLE
Fedora and Ubuntu builds are missing sigc::internal::signal_impl::unreference_exec() symbol

### ISSUE THREAD

**Issue by rathann** [libsigcplusplus/libsigcplusplus#111]
This bug was [originally reported in Fedora](https://bugzilla.redhat.com/show_bug.cgi?id=2321251), but I was asked to report it upstream by Fedora maintainer.

## Description of problem:

Fedora 40 build of libsigc++20 is missing the sigc::internal::signal_impl::unreference_exec() symbol. This causes some third-party proprietary applications (like Horizon Client) to segfault because they refer to and expect that symbol to be present.

## Version-Release number of selected component (if applicable):
`libsigc++20-2.12.1-2.fc40.x86_64`

## How reproducible:
Always

## Steps to Reproduce:
1. Download Debian build from https://packages.debian.org/sid/amd64/libsigc++-2.0-0v5/download
2. Extract the libsigc-2.0.so.0.0.0 shared library
3. `dnf install libsigc++20`
4. `abidiff libsigc-2.0.so.0.0.0 /usr/lib64/libsigc-2.0.so.0.0.0 |c++filt`

## Actual results:
```
Functions changes summary: 0 Removed, 0 Changed, 0 Added function
Variables changes summary: 0 Removed, 0 Changed, 0 Added variable
Function symbols changes summary: 1 Removed, 0 Added function symbol not referenced by debug info
Variable symbols changes summary: 0 Removed, 0 Added variable symbol not referenced by debug info

1 Removed function symbol not referenced by debug info:

  [D] sigc::internal::signal_impl::unreference_exec()
```

## Expected results:
No ABI difference.

## Additional info:
This affects all Fedora releases, including rawhide. Ubuntu package (https://packages.ubuntu.com/noble/libsigc++-2.0-0v5) has the same issue. RHEL9 package libsigc++20-2.10.7-2.el9 has the same issue. Oddly, RHEL8's 2.10.0-6.el8 does have this symbol and a bunch of others, too.

I suspect compiler and/or linker issue. The compiler versions are as follows:

|Distro|GCC|
|------|---|
|Fedora 40, 41 and rawhide|14.2.1|
|Fedora 39|13.3.1|
|RHEL9|11.5.0|
|RHEL8|8.5.0|
|Ubuntu 24.04|14.2.0|
|Debian unstable (Sid)|14.2.0|

There seem to be some differences in how Debian and Fedora build libsigc++20. Debian has this in their [rules](https://salsa.debian.org/gnome-team/libsigcplusplus/-/blob/debian/latest/debian/rules?ref_type=heads) file:
```make
export DEB_BUILD_MAINT_OPTIONS = hardening=+all
export DEB_LDFLAGS_MAINT_APPEND = -Wl,-O1 -Wl,-z,defs

%:
	dh $@ --buildsystem=meson --without autoreconf

override_dh_auto_configure:
	dh_auto_configure -- \
		-Dmaintainer-mode=true \
		-Dwarnings=max \
		-Dbuild-documentation=true
```
while Fedora simply does [this](https://src.fedoraproject.org/rpms/libsigc++20/blob/rawhide/f/libsigc++20.spec#_56):
```spec
%build
%meson -Dbuild-documentation=true
%meson_build
```

**Comment by kjellahl** [comment]
sigc::internal::signal_impl::unreference_exec() is declared `inline`.
Is it possible that inlined functions are included or not included in the
`.so` file depending on compiler or linker options or versions? There are other
`inline` functions. I don't know why only this one would be affected.
When I build libsigc++ 2.12.1 from the tarball at https://download.gnome.org/sources/libsigc++/2.12/,
the sigc::internal::signal_impl::unreference_exec() symbol exists in the `.so` file.
I tried different values of meson's buildtype option (debug, release, minsize and plain).

**Comment by kjellahl** [comment]
From https://packages.ubuntu.com/noble/libsigc++-2.0-0v5 I downloaded the file
libsigc++-2.0_2.12.1-2.debian.tar.xz and extracted it. The extracted file
libsigc++-2.0_2.12.1-2.debian/debian/libsigc++-2.0-0v5.symbols contains this line:
```
 (optional=lto)_ZN4sigc8internal11signal_impl16unreference_execEv@Base 2.4.0
```
(lto = Link-time optimization) Does `(optional=lto)` mean that this symbol can
be omitted?

**Comment by kjellahl** [comment]
> you could maybe file this issue upstream? They might be able to quickly figure out what's going on.

No, I can't quickly (or even slowly) figure out what's going on.
```
abidiff libsigc-2.0.so.0.0.0 /usr/lib64/libsigc-2.0.so.0.0.0 |c++filt

  [D] sigc::internal::signal_impl::unreference_exec()
```
Does this mean that a symbol is removed when the libsigc-2.0.so.0.0.0 file is installed?
Can that be the result of something that has been done in https://github.com/libsigcplusplus/libsigcplusplus?

**Comment by rathann** [comment]
> > you could maybe file this issue upstream? They might be able to quickly figure out what's going on.
> 
> No, I can't quickly (or even slowly) figure out what's going on.
> 
> ```
> abidiff libsigc-2.0.so.0.0.0 /usr/lib64/libsigc-2.0.so.0.0.0 |c++filt
> 
>   [D] sigc::internal::signal_impl::unreference_exec()
> ```
> 
> Does this mean that a symbol is removed when the libsigc-2.0.so.0.0.0 file is installed?

It means that when comparing Debian and Fedora builds, the Fedora one doesn't have this symbol.

> Can that be the result of something that has been done in https://github.com/libsigcplusplus/libsigcplusplus?

I don't know.

**Comment by kjellahl** [comment]
So different distributions have different sets of symbols in their .so files?
And different versions of Fedora have different sets of symbols in their .so files?
I can't explain it. Like I said in my first comment, sigc::internal::signal_impl::unreference_exec()
is declared `inline`. The `inline` keyword is just a hint to the compiler. It's
not forced to inline a function just because the function is declared `inline`.
One compiler with one set of options may perhaps assume that the function will
always be inlined, and it's not necessary to include the function in the .so file.
Another compiler, or the same compiler with a different set of options, may decide otherwise.

The source code of sigc::internal::signal_impl::unreference_exec() is old.
It has not been changed for 20 years. I don't see how I can help here.
You have to pass the question back to Fedora.

**Comment by dhinkeldey2032-netizen** [comment]
jjjjjjjjj
